pkg/protocol/engines.go

```diff
@@ -0,0 +1,261 @@
+package protocol
+
+import (
+	"os"
+	"path/filepath"
+
+	"github.com/google/uuid"
+
+	"github.com/iotaledger/hive.go/ds/reactive"
+	"github.com/iotaledger/hive.go/ierrors"
+	"github.com/iotaledger/hive.go/lo"
+	"github.com/iotaledger/hive.go/runtime/ioutils"
+	"github.com/iotaledger/hive.go/runtime/module"
+	"github.com/iotaledger/hive.go/runtime/workerpool"
+	"github.com/iotaledger/iota-core/pkg/protocol/engine"
+	"github.com/iotaledger/iota-core/pkg/protocol/engine/accounts/accountsledger"
+	"github.com/iotaledger/iota-core/pkg/protocol/engine/blocks"
+	"github.com/iotaledger/iota-core/pkg/protocol/engine/eviction"
+	"github.com/iotaledger/iota-core/pkg/storage"
+	"github.com/iotaledger/iota-core/pkg/storage/utils"
+	iotago "github.com/iotaledger/iota.go/v4"
+)
+
+// Engines is a subcomponent of the protocol that exposes the engines that are managed by the protocol.
+type Engines struct {
+	// Main contains the main engine.
+	Main reactive.Variable[*engine.Engine]
+
+	// protocol contains a reference to the Protocol instance that this component belongs to.
+	protocol *Protocol
+
+	// worker contains the worker pool that is used to process changes to the engine instances asynchronously.
+	worker *workerpool.WorkerPool
+
+	// directory contains the directory that is used to store the engine instances on disk.
+	directory *utils.Directory
+
+	// ReactiveModule embeds a reactive module that provides default API for logging and lifecycle management.
+	*module.ReactiveModule
+}
+
+// newEngines creates a new Engines instance.
+func newEngines(protocol *Protocol) *Engines {
+	e := &Engines{
+		Main:           reactive.NewVariable[*engine.Engine](),
+		ReactiveModule: protocol.NewReactiveSubModule("Engines"),
+		protocol:       protocol,
+		worker:         protocol.Workers.CreatePool("Engines", workerpool.WithWorkerCount(1)),
+		directory:      utils.NewDirectory(protocol.Options.BaseDirectory),
+	}
+
+	protocol.Constructed.OnTrigger(func() {
+		shutdown := lo.Batch(
+			e.syncMainEngineFromMainChain(),
+			e.syncMainEngineInfoFile(),
+			e.injectEngineInstances(),
+		)
+
+		e.Shutdown.OnTrigger(func() {
+			shutdown()
+
+			e.Stopped.Trigger()
+		})
+
+		e.Initialized.Trigger()
+	})
+
+	e.Constructed.Trigger()
+
+	return e
+}
+
+// LoadMainEngine loads the main engine from disk or creates a new one if no engine exists.
+func (e *Engines) LoadMainEngine(snapshotPath string) (*engine.Engine, error) {
```

Comment:
Some of the methods are public and others are private. Why?

Reply:
I just copied the EngineManager as is and simply added the reactive variables to minimize changes for possible merged but since we are close to merge I made all the methods that are not used in external tests private.